src/script/conversation/ConversationRepository.ts

```diff
@@ -310,13 +312,154 @@ export class ConversationRepository {
+
+        break;
+    }
+  };
+
+  /**
+   * For the `federation.delete` event: (Backend A has stopped federating with us)
+      - receive the event from backend
+      - leave the conversations locally that are owned by the backend A which was deleted.
+      - remove the deleted backend A users locally from our own conversations.
+      - insert system message to the affected conversations about federation termination.
+   * @param deletedDomain the domain that stopped federating
+   */
+  private onFederationDelete = async (deletedDomain: string) => {
+    const conversationsToLeave = this.conversationState
+      .conversations()
+      .filter(conversation => conversation.domain === deletedDomain);
+
+    conversationsToLeave.forEach(async conversation => {
+      await this.insertFederationStopSystemMessage(conversation, [deletedDomain]);
+      if (conversation.is1to1()) {
+        conversation.status(ConversationStatus.PAST_MEMBER);
+        return;
+      }
+
+      await this.leaveConversation(conversation, false);
+    });
+
+    const conversationsToRemoveTheirDeletedDomainUsers = this.conversationState
+      .conversations()
+      .filter(conversation => conversation.domain !== deletedDomain);
+
+    conversationsToRemoveTheirDeletedDomainUsers.forEach(async conversation => {
+      const usersToRemove = conversation.allUserEntities().filter(user => user.domain === deletedDomain);
+      if (usersToRemove.length === 0) {
+        return;
+      }
+      await this.insertFederationStopSystemMessage(conversation, [deletedDomain]);
+      await this.removeDeletedFederationUsers(conversation, usersToRemove);
+    });
+  };
+
+  /**
+   * For the `federation.connectionRemoved` event: (Backend A & B stopped federating, user is on C)
+    - receive the event from backend
+    - Identify all conversations that are not owned from A or B domain and that contain users from A and B
+      - remove users from A and B from those conversations
+      - insert system message in those conversations about backend A and B stopping to federate
+    - identify all conversations owned by domain A that contains users from B
+      - remove users from B from those conversations
+      - insert system message in those conversations about backend A and B stopping to federate
+    - Identify all conversations owned by domain B that contains users from A
+      - remove users from A from those conversations
+      - insert system message in those conversations about backend A and B stopping to federate
+   * @param domains The domains that stopped federating with each other
+   */
+  private readonly onFederationConnectionRemove = async (domains: string[]) => {
+    const selfUser = this.userState.self();
+    const [domainOne, domainTwo] = domains;
+    const allConversations = this.conversationState.conversations();
+
+    allConversations
+      .filter(conversation => conversation.domain === domainOne)
```

Comment:
these two blocks are almost identical except domain type. We can create a function and call it twice with each domain.

```
const handleFederationUsers = async (domainToFilter: string, userDomainToDelete: string) => {
  allConversations
    .filter(conversation => conversation.domain === domainToFilter)
    .forEach(async conversation => {
      const usersToDelete = conversation.allUserEntities().filter(user => user.domain === userDomainToDelete);
      if (usersToDelete.length > 0) {
        await this.removeDeletedFederationUsers(conversation, usersToDelete);
        await this.insertFederationStopSystemMessage(conversation, [domainOne, domainTwo]);
      }
    });
};

await handleFederationUsers(domainOne, domainTwo);
await handleFederationUsers(domainTwo, domainOne);
```

Reply:
that's is a good suggestion, indeed. But without unit tests to back it, it's also risky. 
I'd recommend writing tests first (I can help with that), make sure that we are compliant with the specs and then refactoring to a cleaner/more performant code.